domains/client/cross-domain-message-gossip/src/gossip_worker.rs

```diff
@@ -0,0 +1,185 @@
+        let mut incoming_cross_domain_messages = Box::pin(
+            self.gossip_engine
+                .lock()
+                .messages_for(topic::<Block>())
+                .filter_map(|notification| async move {
+                    Message::decode(&mut &notification.message[..]).ok()
+                }),
+        );
+
+        loop {
+            futures::select! {
+                cross_domain_message = incoming_cross_domain_messages.next() => {
+                    if let Some(msg) = cross_domain_message {
+                        tracing::info!(target: LOG_TARGET, "Incoming cross domain message for domain: {:?}", msg.domain_id);
+                        self.handle_cross_domain_message(msg);
+                    }
+                },
+
+                cross_domain_message = gossip_msg_stream.next() => {
+                    if let Some(msg) = cross_domain_message {
+                        tracing::info!(target: LOG_TARGET, "Incoming cross domain message for domain: {:?}", msg.domain_id);
+                        self.handle_cross_domain_message(msg);
+                    }
+                }
+            }
+        }
+    }
+
+    fn handle_cross_domain_message(&mut self, msg: Message) {
+        // mark and rebroadcast message
+        let encoded_msg = msg.encode();
+        self.gossip_validator.note_broadcast(&encoded_msg);
+        self.gossip_engine
+            .lock()
+            .gossip_message(topic::<Block>(), encoded_msg, false);
+
+        let Message {
+            domain_id,
+            encoded_data,
+        } = msg;
+        let sink = match self.domain_tx_pool_sinks.get(&domain_id) {
+            Some(sink) => sink,
+            None => return,
+        };
+
+        // send the message to the open and ready channel
+        if !sink.is_closed() && sink.unbounded_send(encoded_data).is_ok() {
+            return;
+        }
+
+        // sink is either closed or failed to send unbounded message
+        // consider it closed and remove the sink.
+        tracing::error!(
+            target: LOG_TARGET,
+            "Failed to send incoming domain message: {:?}",
+            domain_id
+        );
+        self.domain_tx_pool_sinks.remove(&domain_id);
+    }
+}
+
+/// Gossip validator to retain or clean up Gossiped messages.
+#[derive(Debug, Default)]
+struct GossipValidator {
+    should_broadcast: RwLock<HashSet<MessageHash>>,
+}
+
+impl GossipValidator {
+    fn note_broadcast(&self, msg: &[u8]) {
+        let msg_hash = twox_256(msg);
+        let mut msg_set = self.should_broadcast.write();
+        msg_set.insert(msg_hash);
+    }
+
+    fn should_broadcast(&self, msg: &[u8]) -> bool {
+        let msg_hash = twox_256(msg);
+        let msg_set = self.should_broadcast.read();
+        msg_set.contains(&msg_hash)
+    }
+
+    fn note_broadcasted(&self, msg: &[u8]) {
+        let msg_hash = twox_256(msg);
+        let mut msg_set = self.should_broadcast.write();
+        msg_set.remove(&msg_hash);
+    }
```

Comment:
Should be fine. The underlying Engine is tracking, using lurcache,  the messages it has gossiped and also tracking the peer known as well. So peer should not gossip and even if they did, they would be marked duplicate anyway.

Reply:
Hm... that is also something I was thinking about. Since underlying engine probably already hashes data, it seems like we're wasting resources hashing it again. However I don't see how we can access that pre-hashed value :shr

Reply:
Yeah would have been nice get message hash along with message :(